### DISCUSSION TITLE
OpenAPI validators

### DISCUSSION THREAD

**Discussion by pvdbosch** [belgif/rest-guide#146] [General]
In this discussion, you can share your ideas or experiences about validators that you use to verify if requests and responses comply with the OpenAPI specifications of your REST APIs that follow the REST guide.

**Comment by pvdbosch** [comment]
This thread is meant as a continuation of the discussions on OpenAPI validators in issue #25.

**Comment by pvdbosch** [comment]
I did some tests with the [atlassian swagger-request-validator](https://bitbucket.org/atlassian/swagger-request-validator) to check if it works well with the use of oneOf like we applied it in the [BelgianRemittanceInformation schema](https://github.com/belgif/openapi-money/blob/master/src/main/openapi/money/v1/money-v1.yaml#L71)

It seems to work as expected. 

Error when both structured and unstructured fields are present:
```JSON
{
"messages" : [ {
    "key" : "validation.request.body.schema.oneOf",
    "level" : "ERROR",
    "message" : "Instance failed to match exactly one schema (matched 2 out of 2)",
    "context" : {
      "requestPath" : "/submitRemittanceInformation",
      "apiRequestContentType" : "application/json",
      "location" : "REQUEST",
      "pointers" : {
        "instance" : "/",
        "schema" : "/components/schemas/BelgianRemittanceInformation"
      },
      "requestMethod" : "POST"
    }
  } ]
}
```

Error when none are present:
```JSON
{
  "messages" : [ {
    "key" : "validation.request.body.schema.oneOf",
    "level" : "ERROR",
    "message" : "Instance failed to match exactly one schema (matched 0 out of 2)",
    "additionalInfo" : [ "/oneOf/0: Object has missing required properties ([\"unstructured\"])", "/oneOf/1: Object has missing required properties ([\"structured\"])" ],
    "nestedMessages" : [ {
      "key" : "validation.request.body.schema.required",
      "level" : "ERROR",
      "message" : "Object has missing required properties ([\"unstructured\"])",
      "context" : {
        "pointers" : {
          "instance" : "/",
          "schema" : "/oneOf/0"
        }
      }
    }, {
      "key" : "validation.request.body.schema.required",
      "level" : "ERROR",
      "message" : "Object has missing required properties ([\"structured\"])",
      "context" : {
        "pointers" : {
          "instance" : "/",
          "schema" : "/oneOf/1"
        }
      }
    } ],
    "context" : {
      "requestPath" : "/submitRemittanceInformation",
      "apiRequestContentType" : "application/json",
      "location" : "REQUEST",
      "pointers" : {
        "instance" : "/"
      },
      "requestMethod" : "POST"
    }
  } ]
}
```

**Comment by pvdbosch** [comment]
For the handling of unknown properties in JSON requests, like in [rule-req-valid](https://www.belgif.be/specification/rest/api-guide/#rule-req-valid), I think it might be possible with the workaround of creating separate request and response validators:

```JAVA
var levelResolverResponse = LevelResolver.create().withLevel("validation.schema.additionalProperties", ValidationReport.Level.IGNORE).build();
var levelResolverRequest = LevelResolver.create().withLevel("validation.schema.additionalProperties", ValidationReport.Level.ERROR).build(); // this is actually the (non-spec compliant) default behavior of the validator 
var requestValidator = OpenApiInteractionValidator.createFor(openApiFile)
        .withLevelResolver(levelResolverRequest).build();
var responseValidator = OpenApiInteractionValidator.createFor(openApiFile)
        .withLevelResolver(levelResolverResponse).build();
        
requestValidator.validateRequest(...);
responseValidator.validateResponse(...);       
```

It might be more difficult to configure when using one of the built-in integrations of the validator like the spring one.

**Comment by JDMKSZ** [comment]
In the search for having a validator that validates 1) all examples and 2) that all security scopes are defined in the securitySchemes, I found that

1. Only [openapi-examples-validator](https://www.npmjs.com/package/openapi-examples-validator) has the option to refuse all properties that are not defined in the schema (`-n`). It also has the option to require all properties defined in the schema to be present in the example (`-r`). Typically we don't need the `-r`, because you'd want to create examples with not all properties present (e.g. an example of a resource with "minimal" info present and one with "maximal" info present, or with info A present and one with info B present). However, we DO want the `-n`, to avoid examples becoming outdated when renaming properties (in the API design phase). Usage: `npx openapi-examples-validator -n src/main/openapi/openapi.yaml`. Note that you need to ignore the numerous `"additionalProperties" flag not set` warnings, see [here](https://github.com/codekie/openapi-examples-validator?tab=readme-ov-file#caveat) why.
3. Only [Spectral](https://meta.stoplight.io/docs/spectral/) reports scopes that aren't defined in securitySchemes. The spec says says "In case of OAuth 2, the scopes used in security must be previously defined in securitySchemes.". Spectral has such a validation rule in its default ruleset. Usage: `echo extends: ["spectral:oas", "spectral:asyncapi"] > .spectral.yaml; npx @stoplight/spectral-cli lint src/main/openapi/openapi.yaml --ruleset .spectral.yaml`

These are the tools that I tried:

1. [Redocly CLI](https://github.com/Redocly/redocly-cli)
2. [Smals/Belgif styleguide validation plugin](https://git-cicd.gcloud.belgium.be/rest/styleguide-validation) 1.4 -> examples are only validated w.r.t. the schema, but not the presence of undocumented properties
5. [openapi-style-validator](https://github.com/OpenAPITools/openapi-style-validator)
6. [openapi-examples-validator](https://github.com/codekie/openapi-examples-validator)
7. [Spectral](https://meta.stoplight.io/docs/spectral/)
8. [openapi-linter by superfaceai](https://github.com/superfaceai/openapi-linter)

One more that I found but did not evaluate: [IBM openapi-validator](https://github.com/IBM/openapi-validator)

**Reply by pvdbosch** [reply]
I opened issues on the belgif plugin project for this functionality:
* https://github.com/belgif/rest-guide-validator/issues/3
* https://github.com/belgif/rest-guide-validator/issues/4

**Comment by pvdbosch** [comment]
Detection of unknown properties (for examples or for [rule-req-valid](https://www.belgif.be/specification/rest/api-guide/#rule-req-valid)), should probably be done by injecting `unevaluatedProperties: false` rather than `additionalProperties: false`. 

Injecting `additionalProperties: false`, like openapi-examples-validator and swagger-request-validator do, doesn't work for schema composition (e.g. using `allOf` or `oneOf`), see [Extending Closed Schemas](https://json-schema.org/understanding-json-schema/reference/object#extending).

[unevaluatedProperties](https://json-schema.org/draft/2020-12/json-schema-core#name-unevaluatedproperties) does take nested schemas into account, **but** is only available in OpenAPI >= 3.1 (JSON schema >= 2019-09). It may also cause some unwanted violations (that need to be filtered) when an evaluated property isn't valid, see https://github.com/networknt/json-schema-validator/issues/1100
